Fix DDP join() API in the case of model.no_sync()

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44427 Fix DDP join() API in the case of model.no_sync()**

Closes https://github.com/pytorch/pytorch/issues/44425
DDP join API currently does not work properly with `model.no_sync()`, see https://github.com/pytorch/pytorch/issues/44425 for details. This PR fixes the problem via the approach mentioned in the issue, namely scheduling an allreduce that tells joined ranks whether to sync in the backwards pass or not. Tests are added for skipping gradient synchronization for various `sync_interval`s. 

Differential Revision: [D23609070](https://our.internmc.facebook.com/intern/diff/D23609070/)

rohan-varma committed:
Fix DDP join() API in the case of model.no_sync()

Closes https://github.com/pytorch/pytorch/issues/44425

Differential Revision: [D23609070](https://our.internmc.facebook.com/intern/diff/D23609070/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Fix DDP join() API in the case of model.no_sync()"


Closes https://github.com/pytorch/pytorch/issues/44425
DDP join API currently does not work properly with `model.no_sync()`, see https://github.com/pytorch/pytorch/issues/44425 for details. This PR fixes the problem via the approach mentioned in the issue, namely scheduling an allreduce that tells joined ranks whether to sync in the backwards pass or not. Tests are added for skipping gradient synchronization for various `sync_interval`s. 

Differential Revision: [D23609070](https://our.internmc.facebook.com/intern/diff/D23609070/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Fix DDP join() API in the case of model.no_sync()"


Closes https://github.com/pytorch/pytorch/issues/44425
DDP join API currently does not work properly with `model.no_sync()`, see https://github.com/pytorch/pytorch/issues/44425 for details. This PR fixes the problem via the approach mentioned in the issue, namely scheduling an allreduce that tells joined ranks whether to sync in the backwards pass or not. Tests are added for skipping gradient synchronization for various `sync_interval`s. 

Differential Revision: [D23609070](https://our.internmc.facebook.com/intern/diff/D23609070/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Fix DDP join() API in the case of model.no_sync()"


Closes https://github.com/pytorch/pytorch/issues/44425
DDP join API currently does not work properly with `model.no_sync()`, see https://github.com/pytorch/pytorch/issues/44425 for details. This PR fixes the problem via the approach mentioned in the issue, namely scheduling an allreduce that tells joined ranks whether to sync in the backwards pass or not. Tests are added for skipping gradient synchronization for various `sync_interval`s. 

Differential Revision: [D23609070](https://our.internmc.facebook.com/intern/diff/D23609070/)

[ghstack-poisoned]

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -587,6 +587,10 @@ def forward(self, *inputs, **kwargs):
 
+        if self.ddp_join_enabled:
+            # Notify joined ranks whether they should sync in backwards pass or not.
+            self._check_global_requires_backward_grad_sync(call_from_joiner=False)
```
nit: `s/call_from_joiner/is_joined_rank/g`

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -802,6 +819,19 @@ def join(self, divide_by_initial_world_size=True, enable=True):
+                        work.wait()
+                        # If nonzero, then we should sync in the bwd pass.
+                        should_sync_backwards = should_sync_backwards_tensor.item() != 0
+                        self.require_forward_param_sync = should_sync_backwards
```
Add a comment explaining why we need to set `require_forward_param_sync` here.

rohan-varma committed:
Update on "Fix DDP join() API in the case of model.no_sync()"


Closes https://github.com/pytorch/pytorch/issues/44425
DDP join API currently does not work properly with `model.no_sync()`, see https://github.com/pytorch/pytorch/issues/44425 for details. This PR fixes the problem via the approach mentioned in the issue, namely scheduling an allreduce that tells joined ranks whether to sync in the backwards pass or not. Tests are added for skipping gradient synchronization for various `sync_interval`s. 

Differential Revision: [D23609070](https://our.internmc.facebook.com/intern/diff/D23609070/)

[ghstack-poisoned]
